### DISCUSSION TITLE
Call JavaScript method with C# object as parameter

### DISCUSSION THREAD

**Discussion by achimmihca** [sebastienros/jint#1433] [Q&A] [answered]
The documentation shows how to create a C# object instance from JS:

`engine.SetValue("TheType", TypeReference.CreateTypeReference(engine, typeof(TheType))) // C#`
`var o = new TheType(); // JS`

My question is, how can I pass an instance of `TheType` to a JS method / function?

```
        ObjectInstance jsObject = new Engine()
            .Execute(@"
                class MyJsClass {
                    doSomething(csObj) {
                        return csObj == null;
                    }
             }")
            .Evaluate("new MyJsClass()")
            .AsObject();

        string jsResult = jsObject.Get("doSomething")
            // Call method takes arguments as JsValue array.
            // How to convert the C# object to a JSValue[] ?
            .Call(jsObject, new TheType())
            .AsString();
```

**Comment by lahma** [comment]
Might be easier like this after registering the type to engine.

```c#
const result = engine.Execute("const jsObject = new MyJsClass();").Evalute("jsObject.doSomething(new TheType());");
```
Or
```c#
engine.SetValue("myType", new TheType());
const result = engine.Execute("const jsObject = new MyJsClass();").Evalute("jsObject.doSomething(myType));");
```

**Reply by achimmihca** [reply]
Thanks, `engine.SetValue` got me on the right track. It uses `JsValue.FromObject` internally.

**Comment by achimmihca** [comment, marked as answer]
JsValue.FromObject can be used to convert C# objects to its JS counterpart.

**Reply by achimmihca** [reply]
Sadly, in my test this does not seem to preserve the member values of the C# object.
Instead, a new instance seems to be passed to JS.

I guess what I would need is a serialization from and to JS.

**Reply by lahma** [reply]
Jint creates a wrapper around any non-JS object passed to engine and uses reflection to set values against the original instance.
